src/System.Net.HttpListener/src/System/Net/Managed/HttpListenerRequest.Managed.cs

```diff
@@ -383,32 +380,46 @@ public bool KeepAlive
 
         public Guid RequestTraceIdentifier => Guid.Empty;
 
-        public IAsyncResult BeginGetClientCertificate(AsyncCallback requestCallback, object state)
+        private IAsyncResult BeginGetClientCertificateCore(AsyncCallback requestCallback, object state)
         {
-            Task<X509Certificate2> getClientCertificate = new Task<X509Certificate2>(() => GetClientCertificate());
-            return TaskToApm.Begin(getClientCertificate, requestCallback, state);
+            var asyncResult = new GetClientCertificateAsyncResult(this, state, requestCallback);
+
+            // The certificate is already retrieved by the time this method is called.
+            ClientCertState = ListenerClientCertState.Completed;
+            asyncResult.InvokeCallback(GetClientCertificate());
```

Comment:
> Is it actually fast / non-blocking?

Here's the code:
```cs
        internal X509Certificate2 ClientCertificate
        {
            get { return _clientCert; }
        }
```

Thanks for your detailed explaination - I've never quite got my head around the threading APIs in .NET yet - there seems to be so many ways to do the same thing but with subtle differences - so appreciate it!